### DISCUSSION TITLE
Is there interoperability between WebGPU in JS and wgpu in Rust?

### DISCUSSION THREAD

**Discussion by Gonkalbell** [gfx-rs/wgpu#3821] [Q&A]
I was thinking of trying to convert an existing tutorial from Javascript & WebGPU to Rust, so it could run both on the web and on desktop. Is it possible to convert between things like `wgpu::Device` to and from `GPUDevice` in Javascript, or access the underlying WebGPU objects from wgpu? This way, I'd be able to replace one function at a time, rather than rewrite everything at once or rely on `web_sys`.

From looking through this crate, it seems like any `web_sys` stuff is kept as a private implementation detail, but I could have just missed something. If this isn't possible yet, is there any plan in the future to enable this interop?

**Comment by cwfitzgerald** [comment]
Not currently, but would be good thing to possibly generalize with the existing `as_hal` stuff for accessing underlying apis.

**Reply by bes** [reply]
I have a similar question, we have a video decoding pipeline written in Rust that uses nvidia cards and cuda to decode video frames go opengl textures. I would like to use those textures from wgpu (with a OpenGL target build).

How far off is that? What are the missing pieces?

Thanks 🤗
